core/src/main/java/com/datastrato/gravitino/catalog/TableEventDispatcher.java

```diff
@@ -0,0 +1,132 @@
+/*
+ *  Copyright 2024 Datastrato Pvt Ltd.
+ *  This software is licensed under the Apache License version 2.
+ */
+
+package com.datastrato.gravitino.catalog;
+
+import com.datastrato.gravitino.NameIdentifier;
+import com.datastrato.gravitino.Namespace;
+import com.datastrato.gravitino.exceptions.NoSuchSchemaException;
+import com.datastrato.gravitino.exceptions.NoSuchTableException;
+import com.datastrato.gravitino.exceptions.TableAlreadyExistsException;
+import com.datastrato.gravitino.listener.event.CreateTableEvent;
+import com.datastrato.gravitino.listener.event.CreateTableFailureEvent;
+import com.datastrato.gravitino.listener.event.DropTableEvent;
+import com.datastrato.gravitino.listener.event.DropTableFailureEvent;
+import com.datastrato.gravitino.listener.impl.EventBus;
+import com.datastrato.gravitino.listener.info.TableInfo;
+import com.datastrato.gravitino.rel.Column;
+import com.datastrato.gravitino.rel.Table;
+import com.datastrato.gravitino.rel.TableCatalog;
+import com.datastrato.gravitino.rel.TableChange;
+import com.datastrato.gravitino.rel.expressions.distributions.Distribution;
+import com.datastrato.gravitino.rel.expressions.sorts.SortOrder;
+import com.datastrato.gravitino.rel.expressions.transforms.Transform;
+import com.datastrato.gravitino.rel.indexes.Index;
+import com.datastrato.gravitino.utils.PrincipalUtils;
+import java.util.Map;
+
+/**
+ * {@code TableEventDispatcher} is a decorator for {@link TableCatalog} that not only delegates
+ * table operations to the underlying catalog dispatcher but also dispatches corresponding events to
+ * an {@link EventBus} after each operation is completed. This allows for event-driven workflows or
+ * monitoring of table operations.
+ */
+public class TableEventDispatcher implements TableCatalog {
```

Comment:
It's a bit unintuitive to use `TableCatalog` for dispatcher, literally they're two things.

Maybe we can define an interface like:

```java
interface TableDispatcher extends TableCatalog { }
```

And use this interface for bother `TableOperationDispatcher` and `TableEventDispatcher`. WDYT?

Reply:
yes, this is more general

Reply:
updated, please help to review again